[READY] Allow users to change the options by restarting the server

micbou opened the pull request:
Closes https://github.com/Valloric/YouCompleteMe/issues/3021.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/youcompleteme/3023)
<!-- Reviewable:end -->

bstaletic commented:
Is codecov reporting everything correctly? The only thing that seems no covered is what happens when log level isn't valid.

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[autoload/youcompleteme.vim, line 205 at r1](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LCZJRwP9s88EKhYxAbc:-LCZIyxo9kSSW2X3HrbC:bz98bto) ([raw file](https://github.com/valloric/youcompleteme/blob/b3d99dbbb2871715f4cab3268a15d8c4056ea6f4/autoload/youcompleteme.vim#L205)):*
> ```vim
>   # We need to import ycmd's third_party folders as well since we import and
>   # use ycmd code in the client.
>   su.AddNearestThirdPartyFoldersToSysPath( su.__file__ )
> ```

The old `setup.py` had two `su.AddNearestThirdPartyFoldersToSysPath()` calls. One with `su.__file__` and the other with `DIR_OF_CURRENT_SCRIPT`. Are we missing one call here?

---

*[python/ycm/youcompleteme.py, line 140 at r1](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LCZKHeLCF2tVysvHZa2:-LCZKHeLCF2tVysvHZa3:bhh9y7c) ([raw file](https://github.com/valloric/youcompleteme/blob/b3d99dbbb2871715f4cab3268a15d8c4056ea6f4/python/ycm/youcompleteme.py#L140)):*
> ```Python
>     self._user_options = user_options_store.GetAll()
>     self._omnicomp = OmniCompleter( self._user_options )
>     self._buffers = BufferDict( self._user_options )
> ```

Why set these here instead of in `__init__()`?

---

*[python/ycm/tests/__init__.py, line 52 at r1](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LCZKZkaCeQBsM5WYjph:-LCZKZkaCeQBsM5WYjpi:b-gl2c0n) ([raw file](https://github.com/valloric/youcompleteme/blob/b3d99dbbb2871715f4cab3268a15d8c4056ea6f4/python/ycm/tests/__init__.py#L52)):*
> ```Python
>   'g:ycm_enable_diagnostic_signs': 1,
>   'g:ycm_enable_diagnostic_highlighting': 0,
>   'g:ycm_always_populate_location_list': 0,
> ```

Why do these now need `g:ycm_` prefix?

---


*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023#-:-LCZMQIm8MZUNKmXi6lP:brb96xn)*
<!-- Sent from Reviewable.io -->

micbou commented:
Forgot to update the option of one of the tests. This was causing the unexpected coverage change from codecov.

---

Reviewed 13 of 13 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 205 at r1](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LCZJRwP9s88EKhYxAbc:-LCZPSZh4YmitX2WsLyP:bfm8zcu) ([raw file](https://github.com/valloric/youcompleteme/blob/b3d99dbbb2871715f4cab3268a15d8c4056ea6f4/autoload/youcompleteme.vim#L205)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

The old `setup.py` had two `su.AddNearestThirdPartyFoldersToSysPath()` calls. One with `su.__file__` and the other with `DIR_OF_CURRENT_SCRIPT`. Are we missing one call here?
</blockquote></details>

That call is already covered by `su.AddNearestThirdPartyFoldersToSysPath( script_folder )`. `script_folder` is `YouCompleteMe/autoload` while `DIR_OF_CURRENT_SCRIPT` is `YouCompleteMe/python/ycm`. Both paths share the same `third_party` folder.

---

*[python/ycm/youcompleteme.py, line 140 at r1](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LCZKHeLCF2tVysvHZa2:-LCZQvvxDitW71aZqUtI:b-6ny4p) ([raw file](https://github.com/valloric/youcompleteme/blob/b3d99dbbb2871715f4cab3268a15d8c4056ea6f4/python/ycm/youcompleteme.py#L140)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Why set these here instead of in `__init__()`?
</blockquote></details>

Because they depend on the user options so we need to recreate them when restarting the server.

---

*[python/ycm/tests/__init__.py, line 52 at r1](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LCZKZkaCeQBsM5WYjph:-LCZRCQt2V4B8gqd9Dbu:b-iqi8dl) ([raw file](https://github.com/valloric/youcompleteme/blob/b3d99dbbb2871715f4cab3268a15d8c4056ea6f4/python/ycm/tests/__init__.py#L52)):*
<details><summary><i>Previously, bstaletic (Boris Staletic) wrote…</i></summary><blockquote>

Why do these now need `g:ycm_` prefix?
</blockquote></details>

Because these options are now added to `VIM_OPTIONS` before running a test. `VIM_OPTIONS` is a dictionary holding all Vim options relevant to YCM. Since YCM options start with `g:ycm_`, we need to add the prefix.

---


*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023#-:-LCZQN4vAtfZrZNHKTGz:bthc9py)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
Nice one! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023#-:-LCZaEhsDRRrpMHXSzHe:b6qn8i5)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Reviewed 1 of 13 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 868 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_T5IT6C1ivaEfPJyX:-LC_T5IT6C1ivaEfPJyY:bavprij) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/autoload/youcompleteme.vim#L868)):*
> ```vim
>         \ function( 's:PollServerReady' ) )
> 
>   call s:SetUpOptions()
> ```

shouldn't we do this _before_ calling `RestartServer` ?

It might not be an issue now, but in future we may rely on certain things being set a certain way within the YouCompleteMe instance. Certainly, the previous sequence was calling this prior to starting the server, so it might make sense to call it before restarting the server (if for no other reason than consistency).

---

*[python/ycm/youcompleteme.py, line 110 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_UY3H00Y8Q7UoKKxB:-LC_UY3H00Y8Q7UoKKxC:brqa9vh) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/python/ycm/youcompleteme.py#L110)):*
> ```Python
> class YouCompleteMe( object ):
>   def __init__( self ):
>     self._available_completers = {}
> ```

I wonder if we should blank this out as well in SetUpServer, for consistency.

---

*[python/ycm/tests/__init__.py, line 126 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_WNIwCeJ_UZkKz3k8:-LC_WNIwCeJ_UZkKz3k9:b-v2a1n9) ([raw file](https://github.com/valloric/youcompleteme/blob/acb015fd1c93ffe6c33d81d57955bd437503e7a7/python/ycm/tests/__init__.py#L126)):*
> ```Python
>     @functools.wraps( test )
>     def Wrapper( *args, **kwargs ):
>       ycm = YouCompleteMe( MakeUserOptions( custom_options ) )
> ```

Do we still use MakeUserOptions now ? Fro a quick grep it looks like we replaced all the current usages of it.

---

*[python/ycm/tests/youcompleteme_test.py, line 61 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_XhJy7fHt0JFpQp3e:-LC_XhJy7fHt0JFpQp3f:b-gmpkff) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/python/ycm/tests/youcompleteme_test.py#L61)):*
> ```Python
>   SetVariableValue( 'g:ycm_server_python_interpreter',
>                     '/invalid/path/to/python' )
>   ycm.RestartServer()
> ```

Not a huge deal, but strictly this test is not only testing the restart case, not the initial startup case. If we were to use the options passed to YouCompleteMeInstance it would test the same codepath as before (and wouldn't require the first call to be `StopServer`). Is there value in doing that:

* YouCompleteMeinstance with patched /invalid/python/path
* Then restart server with invalid python path
* then the rest?

Only if you think there's value in covering both cases.

---


*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few minor comments. Thanks for this. Life changed.

---

Reviewed 11 of 13 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023#-:-LC_YMEf1ub3uAjJ_Dup:b-4rik8t)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[autoload/youcompleteme.vim, line 868 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_T5IT6C1ivaEfPJyX:-LC_lCZiCc7PsXSBgF0X:b-dxcbts) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/autoload/youcompleteme.vim#L868)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

shouldn't we do this _before_ calling `RestartServer` ?

It might not be an issue now, but in future we may rely on certain things being set a certain way within the YouCompleteMe instance. Certainly, the previous sequence was calling this prior to starting the server, so it might make sense to call it before restarting the server (if for no other reason than consistency).
</blockquote></details>

Makes sense.

---

*[python/ycm/youcompleteme.py, line 110 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_UY3H00Y8Q7UoKKxB:-LC_c-4F5MRhklMoieBt:bx4l9lv) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/python/ycm/youcompleteme.py#L110)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

I wonder if we should blank this out as well in SetUpServer, for consistency.
</blockquote></details>

It's already the case.

---

*[python/ycm/tests/__init__.py, line 126 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_WNIwCeJ_UZkKz3k8:-LC_lLZ11dh591ix8xaP:blbh83o) ([raw file](https://github.com/valloric/youcompleteme/blob/acb015fd1c93ffe6c33d81d57955bd437503e7a7/python/ycm/tests/__init__.py#L126)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Do we still use MakeUserOptions now ? Fro a quick grep it looks like we replaced all the current usages of it.
</blockquote></details>

I changed it to update the Vim options.

---

*[python/ycm/tests/youcompleteme_test.py, line 61 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_XhJy7fHt0JFpQp3e:-LC_lchc4JscyDa777Hs:b-u54seo) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/python/ycm/tests/youcompleteme_test.py#L61)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Not a huge deal, but strictly this test is not only testing the restart case, not the initial startup case. If we were to use the options passed to YouCompleteMeInstance it would test the same codepath as before (and wouldn't require the first call to be `StopServer`). Is there value in doing that:

* YouCompleteMeinstance with patched /invalid/python/path
* Then restart server with invalid python path
* then the rest?

Only if you think there's value in covering both cases.
</blockquote></details>

Reverted this test and the one below to the initial startup case.

---


*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> @zzbot r+

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[python/ycm/youcompleteme.py, line 110 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_UY3H00Y8Q7UoKKxB:-LC_pCI26byKgZiaGXgn:bdobqml) ([raw file](https://github.com/valloric/youcompleteme/blob/18fc138b167557d39867e8197a4aa087e787aed3/python/ycm/youcompleteme.py#L110)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

It's already the case.
</blockquote></details>

Sorry, meant to delete this comment. Reviewable was confusing me.

---

*[python/ycm/tests/__init__.py, line 126 at r2](https://reviewable.io/reviews/valloric/youcompleteme/3023#-LC_WNIwCeJ_UZkKz3k8:-LC_pWYd1ENoiFgBmvbM:b-7m36bf) ([raw file](https://github.com/valloric/youcompleteme/blob/acb015fd1c93ffe6c33d81d57955bd437503e7a7/python/ycm/tests/__init__.py#L126)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I changed it to update the Vim options.
</blockquote></details>

noice

---


*Comments from [Reviewable](https://reviewable.io/reviews/valloric/youcompleteme/3023#-:-LC_pbM-1GN4fTyidXWE:by5ago3)*
<!-- Sent from Reviewable.io -->
